lib/Semantic_errors.ml

```diff
@@ -0,0 +1,226 @@
+open Core_kernel
+open Pretty_printing
+open Mir
+
+(** Low-level semantic errors used for validation *)
+type t =
+  | IdentifierIsKeyword of location_span * string
+  | IdentifierIsModelName of location_span * string
+  | IdentifierIsStanMathName of location_span * string
+  | IdentifierInUse of location_span * string
+  | IdentifierNotInScope of location_span * string
+  | InvalidIndex of location_span * unsizedtype
+  | TargetOutsideModelBlock of location_span
+  | IllTypedIfReturnTypes of location_span * returntype * returntype
+  | IllTypedTernaryIf of
+      location_span * unsizedtype * unsizedtype * unsizedtype
+  | IllTypedReturningFunction of location_span * string
+  | IllTypedNonReturningFunction of location_span * string
+  | IllTypedStanLibFunApp of location_span * string * unsizedtype list
+  | IllTypedUserFunApp of
+      location_span
+      * string
+      * (autodifftype * unsizedtype) list
+      * returntype
+      * unsizedtype list
+  | IllTypedNotAFunction of location_span * string
+  | IllTypedNotANRFunction of location_span * string
+  | IllTypedNoSuchFunction of location_span * string
+  | IllTypedNoSuchNRFunction of location_span * string
+  | IllTypedBinOp of location_span * Ast.operator * unsizedtype * unsizedtype
+  | IllTypedPrefixOp of location_span * Ast.operator * unsizedtype
+  | IllTypedPostfixOp of location_span * Ast.operator * unsizedtype
+  | FnMapRect of location_span * string
+  | FnConditioning of location_span
+  | FnTargetPlusEquals of location_span
+  | FnRng of location_span
+
+let string_of_operator = mk_string_of Ast.sexp_of_operator
+let ternary_if = "TernaryIf__"
+
+(** A hash table to hold some name conversions between the AST nodes and the
+    Stan Math name of the operator *)
+let string_of_operators =
+  Map.Poly.of_alist_multi
```

Comment:
Is this the best place for this function? Do we not also need something similar to it elsewhere? How does it compare to the functionality in the Operators.ml file?

Reply:
It definitely isn't the right place for it; I think `Operators.ml` is gone on master but I agree that this kind of thing probably lives with the type definition.

Reply:
This functionality still exists in the AST: https://github.com/stan-dev/stanc3/blob/c559488e688c316115650f12d26a2b5e27ed4072/lib/Ast.ml#L292

Reply:
Or no, this is for math function names, that one lives in Semantic_check.ml : https://github.com/stan-dev/stanc3/pull/103/files#diff-3b64248472843a9cb69a14706dd34024R30

Reply:
Yeah, I left it in there too since most of semantic check still uses the exception based approach. If we fully committed to the validation approach it wouldn't be needed since _all_ exception messages would be rendered in 'Semantic_errors.ml' . I can expose from there for the time being which is probably a bit cleaner

Reply:
Ok - I moved the operator related stuff to `Semantic_errors.ml`. Again, the idea would be to have all exception message rendering functions in here and have `Semantic_check.ml` simply construct the low level semantic error data type.